### ISSUE TITLE
HW and distributions in TMB

### ISSUE THREAD

**Issue by noblehendrix** [James-Thorson/2016_Spatio-temporal_models#4]
Hi all,

A question came up after class regarding the HW and how to code the gamma distribution. I just wanted to point out that there are existing probability distributions in TMB that could/should be used, and they are described here:

http://kaskr.github.io/adcomp/group__R__style__distribution.html

Best,
Noble

**Comment by colemonnahan** [comment]
@noblehendrix @James-Thorson  When doing the simulation exercise the intercepts can have different meanings, either due to differences in parameters (e.g., shape in natural space vs expected value in log space) or because of covariates. 

What does it mean then to look at relative errors of this parameter when the simulation and estimation models do not match? Would it be better to find something that is more comparable across models (such as expected value in natural space)?

**Comment by James-Thorson** [comment]
Hmm.  I was imagining that people could compare the log-intercept between
the lognormal and gamma using the following code:

jnll -= dgamma( x, pow(cv,-2), exp(log_intercept)*pow(cv,2), true )

where pow(cv,-2) is the shape parameter, exp(log_intercept)*pow(cv,2) is
the scale parameter, log_intercept is equivalent to the logmean from a
lognormal distribution, and cv is approximately equal to the logsd from a
lognormal distribution.  However, please note that this is the code for the
gamma distribution, and the homework calls for a zero-inflated gamma
distribution.

jim

**Comment by colemonnahan** [comment]
Does not compute.

Why would you use the cv parameter? The EV is shape*scale=exp(log_intercept) regardless of what you have as the scale. Isn't the problem that this is the EV in natural space while the lognormal is in log space?

For a given model could I just calculate the EV as exp(mu+sigma^2/2) for #1 and then scale*shape for #2, regardless of the model matrix?? If I'm thinking about this correctly that would be directly comparable across all of our models. If so, can I just compute the EV (which would be a sum and an integral which is super weird to me) normally?

**Comment by colemonnahan** [comment]
@James-Thorson If you use the 1/CV^2 this is equal to 1/shape. 

`dgamma(x, 1/shape, exp(LP)/shape)`

Then the expected value would be exp(LP)/shape^2. I don't understand what that gets us. 

Can you please clarify what you mean, and how we should compare intercepts when there are covariates? Maybe it makes more sense just to compare within a model?

**Comment by James-Thorson** [comment]
@colemonnahan 

I'm a little confused by your claim.  Algebra shows that shape = 1/CV^2 as follows.  Defining a gamma distribution:

X ~ Gamma( k, Theta )

where k is shape and Theta is scale, then from wikipedia:

E(X) = k * Theta
V(X) = k * Theta^2

Above I claimed without argument that:

k = 1 / CV^2
Theta = mu * CV^2

If this is true, then plugging into the moments from wikipedia:

E(X) = 1/CV^2 * mu*CV^2 = mu
V(X) = 1/CV^2 * mu^2 * CV^4 = mu^2  * CV^2 = sd^2

which checks out (and is the unique solution given two unknowns and two equations).  This confirms that my claim is correct (unless someone wants to point out an error...).

Anyway, for anyone in the class who's reading this, I still encourage you to use the gamma distribution:

jnll -= dgamma( x, pow(cv,-2), exp(log_intercept)*pow(cv,2), true )

Regarding your other point Cole, TMB uses a lognormal distribution:

X ~ Lognormal( logmean, logsd )

and I agree that the expected value for a lognormal is:

E(X) = exp( logmean + logsd^2/2 ) 

which means that mu from the Gamma and logmean from the lognormal are not directly comparable.  However, for the purposes of this homework you're free to compare them.  In a real analysis, you'd want to compare e.g. the expected values directly.

**Comment by colemonnahan** [comment]
@James-Thorson Yeah I had my CV formula inverted for some reason. But my question still stands. As long as the product of the two parameters is exp(LP) then the expected value is exp(LP). I can see how that makes sense biologically.

But what do you gain by reparameterizing the scale? Why not let parameter 1 (p1) be exp(LP)/p2 and p2 be the scale. Then EV[X]=exp(LP) and the interpretation of "scale" remains? Is the point that you use the CV so that you can have mu itself as a parameter? Because it looks like you can't solve for mu and sd^2.

**Comment by James-Thorson** [comment]
perhaps we should talk about this in person, because I'm not following
about "it looks like you can't solve for mu and sd^2".

The point of this re-parameterization is (1) that we have a parameter
representing the expected value (so we can put intepretable covariates on
it), and (2) that the CV is more likely to be constant across different
values of mu than the scale parameter.  This is just an observation, and I
don't offhand have any obvious biological explanation.  Maybe it'd be even
better to have a 2nd parameter representing a constant mean-variance ratio,
like in a Poisson distribution, but I've never seen anyone really discuss
it in fisheries.

**Comment by noblehendrix** [comment]
We should talk about this either before or after class.  There are some interesting aspects of the gamma distribution that would be worth discussing.  Since the CV is constant, the gamma essentially has a tuning parameter for the relationship between mean and variance.  There is also some background on how to construct a gamma GLM that could be worth exploring (canonical link function, etc.).  For example, if you assume the shape (k) is fixed then the variation in the E(y_i) is a function of variability in the scale (Theta_i) parameter and can model covariates B* X_i affecting the scale (Theta_i) parameter, which is more or less Jim's approach (but cleverly disguised by using CV^2).
